happi/audit.py

```diff
@@ -0,0 +1,424 @@
+                           device_class, item)
+            return self.report_code.INVALID
+        else:
+            mod = device_class.rsplit('.')[0]
+        device = item.get('name')
+        if mod in self._all_devices:
+            # has been tested agains PyPi and was found there
+            try:
+                import_class(device_class)
+                return self.report_code.SUCCESS
+            except ImportError as e:
+                logger.warning('For device: %s, %s. Either %s is '
+                               'misspelled or %s is not part of the '
+                               'python environment', device, e,
+                               device_class, device_class)
+                return (self.report_code.INVALID, self.report_code.MISSING)
+        elif mod and mod not in self._all_devices:
+            logger.warning('Provided wrong device/module name: %s. '
+                           'This module does not exist in PyPi', device_class)
+            return self.report_code.INVALID
+
+    def search_pip_package(self, package):
+        """
+        Checks to see if the package is on pypi
+
+        Parameters
+        -----------
+        package: str
+            Name of the package to check
+
+        Returns
+        --------
+            bool
+                To indicate if the package was found or not
+        """
+        process = None
+        try:
+            process = subprocess.Popen('/bin/bash', stdin=subprocess.PIPE,
+                                       stdout=subprocess.PIPE,
+                                       encoding='utf-8')
+            arguments = ['pip', 'search', package]
+            command = ' '.join(arguments)
+
+            out, err = process.communicate(command)
+
+            # the output of 'pip search' will give all the fuzzy searches
+            # we only care if the exact package name is there
+            the_match = re.compile(package)
+            if re.match(the_match, out):
+                return True
+            else:
+                return False
+        except Exception as e:
+            logger.warning(e)
+            if process:
+                process.kill()
+
+    def check_device_in_pypi(self):
+        if self._all_devices:
+            temp_set = self._all_devices.copy()
+            for package in self._all_devices:
+                is_package_found = self.search_pip_package(package)
+                if is_package_found:
+                    logger.info('%s was found in PyPi', package)
+                else:
+                    # maybe a wrong device name was provided because
+                    # it does not exit
+                    logger.info('%s does not exit in PyPi', package)
+                    temp_set.remove(package)
+            # remove the missing package from the set
+            self._all_devices = temp_set
+            return self._all_devices
+
+    def get_value(self, info, item):
```

Comment:
`HappiItem` is a mapping, so it can be dictified `dict(item)[info.key]`

Reply:
thanks a lot for this one.... haha, it bugged me that I could not find a one-liner for it